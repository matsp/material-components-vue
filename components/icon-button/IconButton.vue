<template>
  <a
    v-if="href !== ''"
    :href="href"
    :data-toggle-on-content="toggleOnContent"
    :data-toggle-on-label="toggleOnLabel"
    :data-toggle-on-class="toggleOnClass"
    :data-toggle-off-content="toggleOffContent"
    :data-toggle-off-label="toggleOffLabel"
    :data-toggle-off-class="toggleOffClass"
    v-bind="$attrs"
    class="mdc-icon-button"
    :class="classes"
    v-on="$listeners"
    @MDCIconButtonToggle:change="$emit('change', $event.detail.isOn)"
  >
    <slot v-if="isIconButton" />
    <slot
      v-if="isToggleButtonViaSlots && !value"
      name="toggleOn"
    />
    <slot
      v-if="isToggleButtonViaSlots && value"
      name="toggleOff"
    />
  </a>
  <button
    v-else
    :data-toggle-on-content="toggleOnContent"
    :data-toggle-on-label="toggleOnLabel"
    :data-toggle-on-class="toggleOnClass"
    :data-toggle-off-content="toggleOffContent"
    :data-toggle-off-label="toggleOffLabel"
    :data-toggle-off-class="toggleOffClass"
    v-bind="$attrs"
    class="mdc-icon-button"
    :class="classes"
    v-on="$listeners"
    @MDCIconButtonToggle:change="$emit('change', $event.detail.isOn)"
  >
    <template v-if="this.icon && this.icon !== ''">
      {{ icon }}
    </template>
    <slot v-if="isIconButton" />
    <slot
      v-if="isToggleButtonViaSlots && !value"
      name="toggleOn"
    />
    <slot
      v-if="isToggleButtonViaSlots && value"
      name="toggleOff"
    />
  </button>
</template>

<script>
import { MDCRipple } from '@material/ripple'
import { MDCIconButtonToggle } from '@material/icon-button'

import { baseComponentMixin, themeClassMixin } from '../base'

export default {
  mixins: [baseComponentMixin, themeClassMixin],
  model: {
    prop: 'value',
    event: 'change'
  },
  props: {
    toggleOnContent: {
      type: String,
      default: ''
    },
    toggleOnLabel: {
      type: String,
      default: ''
    },
    toggleOnClass: {
      type: String,
      default: ''
    },
    toggleOffContent: {
      type: String,
      default: ''
    },
    toggleOffLabel: {
      type: String,
      default: ''
    },
    toggleOffClass: {
      type: String,
      default: ''
    },
    value: {
      type: Boolean,
      default: false
    },
    href: {
      type: String,
      default: ''
    },
    icon: {
      type: String,
      default: ''
    }
  },
  data () {
    return {
      mdcIconButtonToggle: undefined,
      mdcRipple: undefined,
      slotObserver: undefined
    }
  },
  computed: {
    isToggleButtonViaSlots () {
      return this.$slots.toggleOn && this.$slots.toggleOff && !this.isIconButton
    },
    isIconButton () {
      return this.$slots.default
    },
    isToggleButton () {
      return this.toggleOnContent !== '' && this.toggleOffContent !== ''
    },
    classes () {
      return {
        'material-icons': this.icon && this.icon !== ''
      }
    }
  },
  watch: {
    value (value) {
      if (typeof this.mdcIconButtonToggle !== 'undefined') {
        this.mdcIconButtonToggle.on = value
      }
    }
  },
  mounted () {
    this.update()
    this.slotObserver = new MutationObserver(() => this.update())
    this.slotObserver.observe(this.$el, {
      childList: true,
      subtree: true
    })
  },
  beforeDestroy () {
    if (typeof this.mdcIconButtonToggle !== 'undefined') {
      this.mdcIconButtonToggle.destroy()
    }
    if (typeof this.mdcRipple !== 'undefined') {
      this.mdcRipple.destroy()
    }
    this.slotObserver.disconnect()
  },
  methods: {
    update () {
      if ((this.isToggleButton || this.isToggleButtonViaSlots) && typeof this.mdcIconButtonToggle === 'undefined') {
        this.mdcIconButtonToggle = MDCIconButtonToggle.attachTo(this.$el)
        this.mdcIconButtonToggle.on = this.value
      }
      if (this.isIconButton && typeof this.mdcRipple === 'undefined') {
        this.mdcRipple = MDCRipple.attachTo(this.$el)
        this.mdcRipple.unbounded = true
      }
    }
  }
}
</script>
