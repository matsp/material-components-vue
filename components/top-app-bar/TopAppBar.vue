<template>
  <header
    :class="classes"
    class="mdc-top-app-bar"
    @MDCTopAppBar:nav="onNavigation()"
  >
    <div class="mdc-top-app-bar__row">
      <section
        v-show="hasStartSection"
        class="mdc-top-app-bar__section mdc-top-app-bar__section--align-start"
      >
        <slot name="navigation" />
        <span
          v-if="title.length > 0"
          class="mdc-top-app-bar__title"
        >
          {{ title }}
        </span>
        <slot name="start" />
      </section>
      <slot />
      <section
        v-show="hasEndSection"
        class="mdc-top-app-bar__section mdc-top-app-bar__section--align-end"
        role="toolbar"
      >
        <slot name="end" />
        <slot name="actions" />
      </section>
    </div>
  </header>
</template>

<script>
import { MDCTopAppBar } from '@material/top-app-bar'
import { MDCComponent } from '@material/base/component'

import { baseComponentMixin, themeClassMixin } from '../base'

export default {
  mixins: [baseComponentMixin, themeClassMixin],
  provide () {
    return { getRipple: this.getRipple }
  },
  props: {
    collapsed: {
      type: Boolean,
      default: false
    },
    short: {
      type: Boolean,
      default: false
    },
    prominent: {
      type: Boolean,
      default: false
    },
    dense: {
      type: Boolean,
      default: false
    },
    fixed: {
      type: Boolean,
      default: false
    },
    title: {
      type: String,
      default: ''
    },
    scrollTarget: {
      type: Object,
      default: () => undefined,
      validator: val => val instanceof EventTarget
    }
  },
  data () {
    return {
      mdcTopAppBar: undefined,
      slotObserver: undefined,
      hasStartSection: this.title.length > 0,
      hasEndSection: true
    }
  },
  computed: {
    classes () {
      return {
        'mdc-top-app-bar--short': this.short && !this.collapsed,
        'mdc-top-app-bar--short-collapsed': this.collapsed && this.short,
        'mdc-top-app-bar--prominent': this.prominent && !this.short,
        'mdc-top-app-bar--dense': this.dense && !this.short,
        'mdc-top-app-bar--fixed': this.fixed && !this.short
      }
    }
  },
  watch: {
    scrollTarget (el) {
      if (this.mdcTopAppBar instanceof MDCComponent && el) {
        this.mdcTopAppBar.setScrollTarget(el)
      }
    },
    short () {
      this.$nextTick(() => {
        this.reInstantiate()
      })
    },
    fixed () {
      this.$nextTick(() => {
        this.reInstantiate()
      })
    },
    collapsed () {
      this.$nextTick(() => {
        this.reInstantiate()
      })
    }
  },
  mounted () {
    this.updateSlots()
    this.slotObserver = new MutationObserver(() => this.updateSlots())
    this.slotObserver.observe(this.$el, {
      childList: true,
      subtree: true
    })
    this.instantiate()
  },
  beforeDestroy () {
    this.slotObserver.disconnect()
    this.mdcTopAppBar.destroy()
  },
  methods: {
    updateSlots () {
      this.hasStartSection = this.$slots.navigation == null || this.$slots.start == null || this.title.length === 0
      this.hasEndSection = this.$slots.end || this.$slots.actions
      if (this.$slots.navigation) {
        this.$slots.navigation.forEach(n => {
          if (n.elm instanceof Element) { n.elm.classList.add('mdc-top-app-bar__navigation-icon') }
        })
      }
      if (this.$slots.actions) {
        this.$slots.actions.forEach(n => {
          if (n.elm instanceof Element) {
            n.elm.classList.add('mdc-top-app-bar__action-item')
          }
        })
      }
      this.$nextTick(() => {
        this.reInstantiate()
      })
    },
    onNavigation () {
      this.$emit('nav')
    },
    instantiate () {
      this.mdcTopAppBar = MDCTopAppBar.attachTo(this.$el)
      if (this.scrollTarget && this.scrollTarget !== window) { this.mdcTopAppBar.setScrollTarget(this.scrollTarget) }
      this.$nextTick(() => {
        if (this.mdcTopAppBar.iconRipples_ instanceof Array) {
          this.mdcTopAppBar.iconRipples_.filter(ripple => ripple instanceof MDCComponent).forEach(ripple => {
            ripple.emit('_init')
          })
        }
      })
    },
    reInstantiate () {
      if (this.mdcTopAppBar instanceof MDCComponent) {
        this.mdcTopAppBar.destroy()
      }
      this.instantiate()
    },
    getRipple (el) {
      if (this.mdcTopAppBar instanceof MDCTopAppBar && this.mdcDialog.iconRipples_ instanceof Array) {
        for (const ripple of this.mdcTopAppBar.iconRipples_) {
          if (ripple instanceof MDCComponent && ripple.root_ === el) return ripple
        }
      }
    }
  }
}
</script>
