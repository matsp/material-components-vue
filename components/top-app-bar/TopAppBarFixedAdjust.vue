<script>
export default {
  props: {
    dense: {
      type: Boolean,
      default: false
    },
    short: {
      type: Boolean,
      default: false
    },
    prominent: {
      type: Boolean,
      default: false
    },
    tag: {
      type: String,
      default: 'div'
    }
  },
  computed: {
    classes () {
      return {
        'mdc-top-app-bar--fixed-adjust': true,
        'mdc-top-app-bar--dense-fixed-adjust': this.dense && !this.short && !this.prominent,
        'mdc-top-app-bar--short-fixed-adjust': this.short,
        'mdc-top-app-bar--prominent-fixed-adjust': this.prominent && !this.short && !this.dense,
        'mdc-top-app-bar--dense-prominent-fixed-adjust': this.dense && this.prominent
      }
    }
  },
  render: function (createElement) {
    return createElement(
      this.tag, // tag name,
      {
        class: this.classes
      },
      this.$slots.default // array of children
    )
  }
}
</script>
