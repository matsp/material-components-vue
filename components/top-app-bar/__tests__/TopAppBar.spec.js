import 'mutationobserver-shim'
import { mount } from '@vue/test-utils'
import TopAppBar from '../TopAppBar.vue'
import IconButton from '../../icon-button/IconButton.vue'
import TopAppBarFixedAdjust from '../TopAppBarFixedAdjust'

describe('Top App Bar', () => {
  it('should mount', () => {
    const wrapper = mount(TopAppBar)
    expect(wrapper.isVueInstance()).toBeTruthy()
    expect(wrapper.vm.$data.mdcTopAppBar).toBeDefined()
  })

  it('should render with no prop', () => {
    const wrapper = mount(TopAppBar)
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-top-app-bar')
  })

  it('should render as short', () => {
    const wrapper = mount(TopAppBar, {
      propsData: {
        short: true
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-top-app-bar--short')
  })

  it('should render as short and collapsed', () => {
    const wrapper = mount(TopAppBar, {
      propsData: {
        short: true,
        collapsed: true
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-top-app-bar--short-collapsed')
  })

  it('should render as prominent', () => {
    const wrapper = mount(TopAppBar, {
      propsData: {
        prominent: true
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-top-app-bar--prominent')
  })

  it('should render as dense', () => {
    const wrapper = mount(TopAppBar, {
      propsData: {
        dense: true
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-top-app-bar--dense')
  })

  it('should render as fixed', () => {
    const wrapper = mount(TopAppBar, {
      propsData: {
        fixed: true
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-top-app-bar--fixed')
  })

  it('should render with title', () => {
    const wrapper = mount(TopAppBar, {
      propsData: {
        title: 'title text'
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find('.mdc-top-app-bar__title').text()).toEqual('title text')
  })

  it('should render with optional action button', () => {
    const wrapper = mount(TopAppBar, {
      slots: {
        actions: IconButton
      }
    })
    expect(wrapper.find('.mdc-top-app-bar__action-item').exists()).toBeTruthy()
  })

  it('should render with optional navigation button', () => {
    const wrapper = mount(TopAppBar, {
      slots: {
        navigation: IconButton
      }
    })
    expect(wrapper.find('.mdc-top-app-bar__navigation-icon').exists()).toBeTruthy()
    wrapper.find('.mdc-top-app-bar__navigation-icon').trigger('click')
    expect(wrapper.emitted().nav).toBeTruthy()
  })
})

describe('Top App Bar Fixed Adjust', () => {
  it('should mount', () => {
    const wrapper = mount(TopAppBarFixedAdjust)
    expect(wrapper.isVueInstance()).toBeTruthy()
  })

  it('should render with no prop', () => {
    const wrapper = mount(TopAppBarFixedAdjust)
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-top-app-bar--fixed-adjust')
    expect(wrapper.is('div')).toBe(true)
  })

  it('should render as short', () => {
    const wrapper = mount(TopAppBarFixedAdjust, {
      propsData: {
        short: true
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-top-app-bar--short-fixed-adjust')
  })

  it('should render as dense and prominent', () => {
    const wrapper = mount(TopAppBarFixedAdjust, {
      propsData: {
        dense: true,
        prominent: true
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-top-app-bar--dense-prominent-fixed-adjust')
  })

  it('should render as prominent', () => {
    const wrapper = mount(TopAppBarFixedAdjust, {
      propsData: {
        prominent: true
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-top-app-bar--prominent-fixed-adjust')
  })

  it('should render as dense', () => {
    const wrapper = mount(TopAppBarFixedAdjust, {
      propsData: {
        dense: true
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-top-app-bar--dense-fixed-adjust')
  })

  it('should render as main', () => {
    const wrapper = mount(TopAppBarFixedAdjust, {
      propsData: {
        tag: 'main'
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.is('main')).toBe(true)
  })
})
