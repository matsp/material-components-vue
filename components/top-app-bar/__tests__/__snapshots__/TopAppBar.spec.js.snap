// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Top App Bar Fixed Adjust should render as dense 1`] = `<div class="mdc-top-app-bar--fixed-adjust mdc-top-app-bar--dense-fixed-adjust"></div>`;

exports[`Top App Bar Fixed Adjust should render as dense and prominent 1`] = `<div class="mdc-top-app-bar--fixed-adjust mdc-top-app-bar--dense-prominent-fixed-adjust"></div>`;

exports[`Top App Bar Fixed Adjust should render as main 1`] = `<main class="mdc-top-app-bar--fixed-adjust"></main>`;

exports[`Top App Bar Fixed Adjust should render as prominent 1`] = `<div class="mdc-top-app-bar--fixed-adjust mdc-top-app-bar--prominent-fixed-adjust"></div>`;

exports[`Top App Bar Fixed Adjust should render as short 1`] = `<div class="mdc-top-app-bar--fixed-adjust mdc-top-app-bar--short-fixed-adjust"></div>`;

exports[`Top App Bar Fixed Adjust should render with no prop 1`] = `<div class="mdc-top-app-bar--fixed-adjust"></div>`;

exports[`Top App Bar should render as dense 1`] = `
<header class="mdc-top-app-bar mdc-top-app-bar--dense">
  <div class="mdc-top-app-bar__row">
    <section class="mdc-top-app-bar__section mdc-top-app-bar__section--align-start" style="display: none;">
      <!---->
    </section>
    <section role="toolbar" class="mdc-top-app-bar__section mdc-top-app-bar__section--align-end"> </section>
  </div>
</header>
`;

exports[`Top App Bar should render as fixed 1`] = `
<header class="mdc-top-app-bar mdc-top-app-bar--fixed">
  <div class="mdc-top-app-bar__row">
    <section class="mdc-top-app-bar__section mdc-top-app-bar__section--align-start" style="display: none;">
      <!---->
    </section>
    <section role="toolbar" class="mdc-top-app-bar__section mdc-top-app-bar__section--align-end"> </section>
  </div>
</header>
`;

exports[`Top App Bar should render as prominent 1`] = `
<header class="mdc-top-app-bar mdc-top-app-bar--prominent">
  <div class="mdc-top-app-bar__row">
    <section class="mdc-top-app-bar__section mdc-top-app-bar__section--align-start" style="display: none;">
      <!---->
    </section>
    <section role="toolbar" class="mdc-top-app-bar__section mdc-top-app-bar__section--align-end"> </section>
  </div>
</header>
`;

exports[`Top App Bar should render as short 1`] = `
<header class="mdc-top-app-bar mdc-top-app-bar--short">
  <div class="mdc-top-app-bar__row">
    <section class="mdc-top-app-bar__section mdc-top-app-bar__section--align-start" style="display: none;">
      <!---->
    </section>
    <section role="toolbar" class="mdc-top-app-bar__section mdc-top-app-bar__section--align-end"> </section>
  </div>
</header>
`;

exports[`Top App Bar should render as short and collapsed 1`] = `
<header class="mdc-top-app-bar mdc-top-app-bar--short-collapsed">
  <div class="mdc-top-app-bar__row">
    <section class="mdc-top-app-bar__section mdc-top-app-bar__section--align-start" style="display: none;">
      <!---->
    </section>
    <section role="toolbar" class="mdc-top-app-bar__section mdc-top-app-bar__section--align-end"> </section>
  </div>
</header>
`;

exports[`Top App Bar should render with no prop 1`] = `
<header class="mdc-top-app-bar">
  <div class="mdc-top-app-bar__row">
    <section class="mdc-top-app-bar__section mdc-top-app-bar__section--align-start" style="display: none;">
      <!---->
    </section>
    <section role="toolbar" class="mdc-top-app-bar__section mdc-top-app-bar__section--align-end"> </section>
  </div>
</header>
`;

exports[`Top App Bar should render with title 1`] = `
<header class="mdc-top-app-bar">
  <div class="mdc-top-app-bar__row">
    <section class="mdc-top-app-bar__section mdc-top-app-bar__section--align-start"> <span class="mdc-top-app-bar__title">
        title text
      </span> </section>
    <section role="toolbar" class="mdc-top-app-bar__section mdc-top-app-bar__section--align-end"> </section>
  </div>
</header>
`;
